File: src/runtime/executor.ts
/**
 * Ruby VM (Worker) へのコード送信と実行結果の購読を担当する
 */
export interface RubyVMController {
  run(code: string): void;
}

export class Executor {
  private controller: RubyVMController;

  constructor(controller: RubyVMController) {
    this.controller = controller;
  }

  /**
   * コードを実行する
   * @param {string} code 
   */
  execute(code: string): void {
    if (!code) return;
    this.controller.run(code);
  }
}

File: src/core/ruby/executor.rb
require "stringio"

module Executor
  def self.run(code)
    $stdin = StringIO.new($rubox_stdin || "")
    output = StringIO.new
    $stdout = $stderr = output
    begin 
      sandbox_class = Class.new
      sandbox_class.class_eval(<<~RUBY)
        def get_binding
          binding
        end
      RUBY
      sandbox_class.new.get_binding.eval(code)
    rescue Exception => e
      warn e.full_message
    ensure
      $stdout = STDOUT
      $stderr = STDERR
    end
    output.string
  end
end
